### DISCUSSION TITLE
TM1637 display with ESP32 board is very slow: bug in DelayHelper.DelayMicroseconds

### DISCUSSION THREAD

**Discussion by Fido789** [nanoframework/Home#1246] [General]
Hello!

I am trying to create a simple clock using ESP32 board with TM1637 four character display (based on https://github.com/nanoframework/nanoFramework.IoT.Device/blob/develop/devices/Tm1637/README.md). It works, but no matter what I do the refresh rate for the display is like once per two or three seconds, so totally unusable for my purpose.

Even if I do simple loop displaying 9 4 digit numbers (1111, 2222 to 9999) it is slow, because the change to each number takes like a second. And the digits are changed one by one from left to right, even if in code I set them all at once.

Is there anything I can do to make the refresh rate faster?

**Comment by Ellerbach** [comment]
Bug sits in `DelayHelper.DelayMicroseconds`
https://github.com/nanoframework/nanoFramework.IoT.Device/blob/bcf7847bdbc54c92d3100cb3f0b0fc2d36889413/devices/System.Diagnostics.Stopwatch/DelayHelper.cs#L64

**Comment by josesimoes** [comment]
With the following code running on an ESP32 WROVER KIT
`DelayHelper.DelayMicroseconds(1000_000, true);`

this is what can measure:
![image](https://user-images.githubusercontent.com/1881520/226420232-c8f3e6e8-fa7d-4208-94d4-33777b55bd94.png)

with 
`DelayHelper.DelayMilliseconds(1000, true);`

![image](https://user-images.githubusercontent.com/1881520/226420533-4509a635-d7ae-4222-9e9f-3c7ac3267eef.png)

cross-checking with a diferent number:
`DelayHelper.DelayMilliseconds(25, true);`

![image](https://user-images.githubusercontent.com/1881520/226420992-30f22a81-a3e0-4b5c-81c5-63eff0d004f0.png)

(which is expected from thread slice being 20ms)


And finally:
`DelayHelper.DelayMicroseconds(1, false);`

![image](https://user-images.githubusercontent.com/1881520/226421431-58aa37f2-3cd3-48c5-be36-ddd3531a89c7.png)

Conclusion: DelayMicroseconds doesn't seem to have any bug...

**Reply by Ellerbach** [reply]
> Conclusion: DelayMicroseconds doesn't seem to have any bug...

Thanks for the very detailed investigation. There is then something else with delays that is off. It can't be that slow because there is absolutely nothing else that takes time except potentially this one.
I can't investigate more now. I don't have the proper hardware for that.

**Comment by Ellerbach** [comment]
More elements:
* DelayHelper.DelayMicroseconds(1, false); is actually taking too long.
* For such delay spinwait should be used
* So code change and adjustments need to be done
* And solution tested

**Comment by Ellerbach** [comment]
@Fido789 can you check if the fix I did is working? https://github.com/nanoframework/nanoFramework.IoT.Device/pull/753
I don't have access to any hardware right now.

**Comment by Fido789** [comment]
@Ellerbach Oh yes, that fixes the problem, amazing! Thank you very much for your effort!

**Comment by Ellerbach** [comment]
Fixed in the related PR https://github.com/nanoframework/nanoFramework.IoT.Device/pull/753
